cli/flox-watchdog/src/process.rs

```diff
@@ -0,0 +1,139 @@
+const PROCFS_STATE_ZOMBIE: &str = "Z";
+/// Process is dead and will transition to a zombie or disappear.
+/// Technically we shouldn't see this, but just in case:
+/// https://unix.stackexchange.com/a/653370
+const PROCFS_STATE_DEAD: &str = "X";
+const PROCFS_STATE_DEAD_COMPAT: &str = "x";
+
+/// Process watcher for Linux based on `procfs`.
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct ProcfsWatcher {
+    pid: i32,
+}
+
+/// Allow compilation when not on Linux to make IDE/LSP development easier. It
+/// will always return negative results when not used on other platforms.
+#[cfg_attr(not(target_os = "linux"), allow(dead_code))]
+impl ProcfsWatcher {
+    /// The PID will *not* be checked on instantiation so that we don't need to
+    /// do additional cleanup checks before `wait_for_termination`.
+    pub(crate) fn new(pid: i32) -> Self {
+        Self { pid }
+    }
+
+    /// Reads the state of a process. Any failure is used as an indication that
+    /// the process is no longer running.
+    fn state(&self) -> Option<String> {
+        let path = format!("/proc/{}/stat", self.pid);
+        let stat = match read_to_string(path) {
+            Ok(stat) => stat,
+            Err(err) => {
+                debug!(%err, self.pid, "failed to read stat, treating as dead");
+                return None;
+            },
+        };
+        let state = match stat.split_whitespace().nth(2) {
+            Some(state) => state.to_string(),
+            None => {
+                debug!(self.pid, "failed to parse stat, treating as dead");
+                return None;
+            },
+        };
+        Some(state)
+    }
+
+    /// Returns whether the process is considered running.
+    pub(crate) fn is_running(&self) -> bool {
+        self.state().is_some_and(|state| {
+            !matches!(
+                state.as_str(),
+                PROCFS_STATE_ZOMBIE | PROCFS_STATE_DEAD | PROCFS_STATE_DEAD_COMPAT
+            )
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::process::Command;
+    use std::thread;
+    use std::time::Duration;
+
+    use super::*;
+
+    /// Wait some attempts for the process to reach the desired state
+    fn poll_until_state(watcher: &ProcfsWatcher, state: &str) {
+        for _ in 0..10 {
+            if let Some(s) = watcher.state() {
+                if s == state {
+                    break;
+                }
+            }
+            thread::sleep(Duration::from_millis(100));
+        }
+    }
+
+    #[test]
+    #[cfg_attr(not(target_os = "macos"), ignore)]
+    fn test_procfswatcher_macos_always_reports_false() {
+        let watcher = ProcfsWatcher::new(1);
+        assert_eq!(watcher.is_running(), false);
+    }
+
+    #[test]
+    #[cfg_attr(not(target_os = "linux"), ignore)]
+    fn test_procfswatcher_pid1_always_running() {
+        let watcher = ProcfsWatcher::new(1);
+        assert_eq!(watcher.is_running(), true);
```

Comment:
nit: here and elsewhere you can just use `assert!()` per `cargo clippy --tests`

Reply:
The difference between `assert!(watcher.is_running())` and `assert!(!watcher.is_running())` becomes very subtle and easy to overlook.